### PR TITLE
[HOPSWORKS-640] Flink/Beam

### PR DESCRIPTION
**Pull request by tkakantousis**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [ ] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/yarn/YarnRunner.java [hunk 1]
```diff
@@ -274,76 +268,44 @@ ApplicationId startAppMaster(YarnClientService ycs, String dfsUsername,
...
+        logger.log(Level.FINE, "Deploying Flink cluster.");
+        clusterClient = yarnClusterDescriptor.deploySessionCluster(clusterSpecification);
+        clusterClient.setDetached(true);
+        clusterClient.waitForClusterToBeReady();
+    
+        appId = clusterClient.getClusterId();
+        logger.log(Level.FINE, "Deployed Flink cluster with ID {0}",appId.toString());
+    
+      } catch (ClusterDeploymentException ex) {
+        // TODO: Ahmad handle this exception
```
Comment (ErmiasG) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/yarn/YarnRunner.java]: When is Ahmad going to handle this?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/flink/FlinkCleaner.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+      String archiveDir = flinkController.getArchiveDir();
+      //Delete all without hdfs user
+      dfso = dfs.getDfsOps();
+      List<Inode> jobs = inodeFacade.getChildren(archiveDir);
+      for (Inode job : jobs) {
+        if (job.getHdfsUser() == null) {
+          dfso.rm(new Path(archiveDir + File.separator + job.getInodePK().getName()), false);
+        }
+      }
+    } catch (IOException e) {
```
Comment (ErmiasG) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/flink/FlinkCleaner.java]: Can this catch exception? to make sure it does not get expunged.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/util/SparkConfigurationUtil.java [hunk 3]
```diff
@@ -161,6 +161,8 @@
             HopsUtils.IGNORE);
     addToSparkEnvironment(sparkProps, "HOPSWORKS_PROJECT_ID", Integer.toString(project.getId()),
             HopsUtils.IGNORE);
+    addToSparkEnvironment(sparkProps, "FLINK_CONF_DIR", settings.getFlinkConfDir(),
```
Comment (robzor92) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/util/SparkConfigurationUtil.java]: Why does this need to be set for Spark jobs?

File: hopsworks-IT/pom.xml [hunk 4]
```diff
@@ -69,6 +69,7 @@
     <!--suppress UnresolvedMavenProperty -->
     <glassfish.admin_port>${env.GLASSFISH_ADMIN_PORT}</glassfish.admin_port>
     <glassfish.domain>domain1</glassfish.domain>
+    <maven.deploy.skip>true</maven.deploy.skip>
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-IT/pom.xml]: When doing `mvn deploy` to push hopsworks-common on our maven repo

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/kibana/KibanaProxyServlet.java [hunk 5]
```diff
@@ -100,16 +100,18 @@
...
   };
 
   /**
    * Authorize user to access particular index.
    *
-   * @param servletRequest
-   * @param servletResponse
-   * @throws ServletException
-   * @throws IOException
+   * @param servletRequest servletRequest
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/kibana/KibanaProxyServlet.java]: It's not duplicate. Javadoc spec expects a description there, I was too lazy to write a proper description so I added the variable name

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/elastic/ElasticCleaner.java [hunk 6]
```diff
@@ -63,7 +63,7 @@
   /**
    * Periodically remove job/notebooks log indexes. Runs once per day and removed indices older than 7 days.
    *
-   * @param timer
+   * @param timer timer
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/elastic/ElasticCleaner.java]: See comment above regarding javadoc

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/flink/FlinkCleaner.java (hunk 2 above)
Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/flink/FlinkCleaner.java]: Will change to Exception

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/yarn/YarnExecutionFinalizer.java [hunk 7]
```diff
@@ -174,8 +173,6 @@ public void removeAllNecessary(Execution exec) throws IOException {
     String appDir = "hdfs://" + settings.getHdfsTmpCertDir() + "/" + exec.getHdfsUser() + File.separator + exec.
```
[the remark this answers is not in the document]
Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/yarn/YarnExecutionFinalizer.java]: That line was committed 2 years ago I prefer to have a JIRA to refactor all these "/"

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/util/SparkConfigurationUtil.java (hunk 3 above)
Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/util/SparkConfigurationUtil.java]: Because we run Beam code within PySpark and I need this property in PySpark notebooks.

File: hopsworks-common/pom.xml [hunk 8]
```diff
@@ -355,6 +341,11 @@
       <artifactId>bijection-avro_2.12</artifactId>
       <version>0.9.6</version>
     </dependency>
+    <dependency>
+      <groupId>org.yaml</groupId>
```
Comment (robzor92) [inline hopsworks-common/pom.xml]: Why is this needed?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/spark/SparkJobConfiguration.java [hunk 9]
```diff
@@ -267,6 +267,18 @@ public JobType getJobType() {
       return JobType.SPARK;
     }
   }
+  
+  @Override
+  @XmlElement(name="jobTypeName")
```
Comment (robzor92) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/spark/SparkJobConfiguration.java]: What's the difference between JobType and JobTypeName?

Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/spark/SparkJobConfiguration.java]: JobTypeName is the "display name" of the type. Because a type can be `beam_flink_k8s`, but it's ugly for the client. We could argue that the display name is the responsibility of the client, but in this case this field is not persisted and our front-end is way too spaghetti to have another lookup table.

File: hopsworks-common/pom.xml (hunk 8 above)
Comment (tkakantousis) [reply hopsworks-common/pom.xml]: https://github.com/logicalclocks/hopsworks/pull/299/files#diff-bb2605f3192c961ef2b7393814fcf244R225
